pkg/controller/controller.go

```diff
@@ -0,0 +1,252 @@
+	c.handlePodObject(obj, "add")
+}
+func (c *Controller) handlePodUpdate(obj interface{}) {
+	c.handlePodObject(obj, "update")
+}
+func (c *Controller) handlePodDelete(obj interface{}) {
+	c.handlePodObject(obj, "delete")
+}
+
+func (c *Controller) observePodCreate(pvcKey string) {
+	c.podExpectations.CreationObserved(pvcKey)
+}
+
+func (c *Controller) handlePodObject(obj interface{}, verb string) {
+	var object metav1.Object
+	var ok bool
+	if object, ok = obj.(metav1.Object); !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			runtime.HandleError(errors.Errorf("error decoding object, invalid type"))
+			return
+		}
+		object, ok = tombstone.Obj.(metav1.Object)
+		if !ok {
+			runtime.HandleError(errors.Errorf("error decoding object tombstone, invalid type"))
+			return
+		}
+		glog.V(3).Infof("Recovered deleted object '%s' from tombstone", object.GetName())
+	}
+	glog.V(3).Infof("Processing object: %s", object.GetName())
+	if ownerRef := metav1.GetControllerOf(object); ownerRef != nil {
+		_, createdByUs := object.GetAnnotations()[AnnCreatedBy]
+
+		if ownerRef.Kind != "PersistentVolumeClaim" {
+			return
+		} else if !createdByUs {
+			return
+		}
+
+		var err error
+		var pvc *v1.PersistentVolumeClaim
+		_, annTargetPodNamespace := object.GetAnnotations()[AnnTargetPodNamespace]
+		if annTargetPodNamespace {
+			pvc, err = c.pvcLister.PersistentVolumeClaims(object.GetAnnotations()[AnnTargetPodNamespace]).Get(ownerRef.Name)
+		} else {
+			pvc, err = c.pvcLister.PersistentVolumeClaims(object.GetNamespace()).Get(ownerRef.Name)
+		}
+
+		if err != nil {
+			glog.V(3).Infof("ignoring orphaned object '%s' of pvc '%s'", object.GetSelfLink(), ownerRef.Name)
+			return
+		}
+
+		if verb == "add" {
+			pvcKey, err := cache.MetaNamespaceKeyFunc(pvc)
+			if err != nil {
+				runtime.HandleError(err)
+				return
+			}
+
+			c.observePodCreate(pvcKey)
+		}
+		c.enqueuePVC(pvc)
+		return
+	}
+}
+
+func (c *Controller) enqueuePVC(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited(key)
+}
+
+//Run is being called from cdi controllers
+func (c *Controller) run(threadiness int, stopCh <-chan struct{}, controller interface{}) error { //*CloneContorler
+	defer func() {
+		c.queue.ShutDown()
+	}()
+	glog.V(3).Infoln("Starting cdi controller Run loop")
+	if threadiness < 1 {
+		return errors.Errorf("expected >0 threads, got %d", threadiness)
+	}
+
+	if !cache.WaitForCacheSync(stopCh, c.pvcInformer.HasSynced) {
+		return errors.New("Timeout waiting for pvc cache sync")
+	}
+	if !cache.WaitForCacheSync(stopCh, c.podInformer.HasSynced) {
+		return errors.New("Timeout waiting for pod cache sync")
+	}
+	glog.V(3).Infoln("Controller cache has synced")
+	for i := 0; i < threadiness; i++ {
+		//Go is not pure object oriented language. The command repetition below is a result of that.
+		switch t := controller.(type) {
```

Comment:
Am I losing it, or does this switch do nothing?
Wouldn't this be the same?
```go
t := controller.(type)
go wait.Until(t.runPVCWorkers, time.Second, stopCh)
```
I guess if the type isn't one of the two, but is that even possible?

Reply:
Go does not let me do this... Spent half day to find a solution.
You cannot do what you wrote above as you get an error:
Use of .(type) outside type switch

Reply:
Oh I see, I missed the fact controller was an interface.